### DISCUSSION TITLE
Loop over localized collections not working as expected

### DISCUSSION THREAD

**Discussion by madsem** [11ty/eleventy#1875] [Q&A] [answered]
I'm trying to display collections on my homepage, in the current pages locale.
If I simply do `collections.all` I get output, but in all languages.

Been trying various things but nothing I try actually even displayed anything...

```js
module.exports = {
  layout: "page",
  tags: ["products_{{ locale }}"],
  permalink: "{{ locale }}/{{ slug or page.fileSlug }}/",
};
```

I'm trying to create Collections by localizing the tag associated with each collection.
This is then the code I have on my homepage atm:
```nunjucks
<div class="content flow">

        {% set posts = collections["products_" + locale] %}
        {% for post in posts %}
        {{ locale }}
        <strong>{{post.data.title}}</strong>
        {% endfor %}

    </div>
```

The above leads to no output at all, whereas when I remove the locale from the tags in my datafile, and also from the loop, the following is output (which is also expected):
```html
// english homepage
<div class="content flow">en <strong>Cool List</strong> en <strong>Coole Liste</strong></div>

// german homepage
<div class="content flow">de <strong>Cool List</strong> de <strong>Coole Liste</strong></div>
```

This shows that the locale setup works, but my brain not so much :D

Expected result would be though:
```html
// english homepage
<div class="content flow">en <strong>Cool List</strong></div>

// german homepage
<div class="content flow">de <strong>Coole Liste</strong></div>
```

Ideally, what I would like to achieve is to re-use all my layouts and have 11ty automatically spit out the content only for the currently processed  language.

**Comment by pdehaan** [comment, marked as answer]
I have 3 very rough debugging ideas.

But first, a couple little filters:

```js
eleventyConfig.addFilter("inspect", require("util").inspect);
eleventyConfig.addFilter("keys", (obj) => Object.keys(obj).sort());
```

1. Create a new page and add something like `{{ collections | keys | inspect }}`. That should hopefully tell us which collection names are currently created. My guess is that there isn't a `collections.products_en` created.
2. Put `{{ tags | inspect }}` on one of your pages to make sure that it's doing what you think it's doing.
3. Put `{{ posts | inspect }}` right after `{% set posts = collections["products_" + locale] %}` and confirm that it returns an array of some products. I'm guessing the value is an empty array or some falsy value, which I have some therories on.

**Comment by madsem** [comment]
Thanks @pdehaan 

I've done all the above on my `home.njk` layout, with this code:
```nunjucks
{% set posts = collections["products_" + locale] %}
        {{ posts | inspect }}
        {% for post in posts %}
        <strong>{{post.data.title}}</strong>
        {% endfor %}
        

        {{ collections | keys | inspect }}

        {{ tags | inspect }}
```

And this is the output on the page:
```log
undefined [ 'all', 'pages', 'products_{{ locale }}' ] undefined
```

I would have expected at least some other tags to exist, because i'm also setting tags in other pages:
```js
module.exports = {
  layout: "page",
  tags: ["pages"],
  permalink: "{{ locale }}/{{ page.fileSlug }}/",
};
```

But in regards to the localization, this is indeed all undefined and / or not substituted.

**Reply by pdehaan** [reply]
Yeah, I think we're seeing the same thing:

```sh
tree src

src/
└── products/
    ├── de/
    │   ├── de.11tydata.js
    │   └── hello_world.njk
    ├── en/
    └── products.11tydata.js

3 directories, 3 files
```

My src/products/products.11tydata.js looks like this:
```js
module.exports = {
  // layout: "page",
  tags: ["products_{{ locale }}"],
  permalink: "{{ locale }}/{{ slug or page.fileSlug }}/",
};
```

My src/products/de/hello_world.njk looks like this:
```njk
---
title: Sample Page!
---

locale={{ locale }}
tags={{ tags }}
collections={{ collections | keys | inspect | safe }}
```

### OUTPUT

```ini
locale=de
tags=products_{{ locale }}
collections=[ 'all', 'products_{{ locale }}' ]
```

So that is telling me that the `tags` key iin the products.11tydata.js file doesn't substitute the `locale` (which isn't entirely surprising&hellip; `permalink` is special, I think for everything else you have to use [`eleventyComputed`](https://www.11ty.dev/docs/data-computed/)).

---

Now if I switch my products.11tydata.js file to use `eleventyComputed` I'm seeing the correct tag values, but...

```js
module.exports = {
  "permalink": "{{ locale }}/{{ slug or page.fileSlug }}/",
  eleventyComputed: {
    tags(data) {
      return [`products_${ data.locale }`];
    }
  }
};
```

### OUTPUT

```ini
locale=de
tags=products_de
collections=[ 'all' ]
```

... but we see that adding dynamic computed tags doesn't automatically add them into a collection. There should be a bunch of issues and potential solutions in the GitHub issues, but I can't immediately think of something that magically worked in every scenario. Not sure if you could manually create the collections via .eleventy.js and [`.addCollection()`](https://www.11ty.dev/docs/collections/#advanced-custom-filtering-and-sorting) since you have a predictable file structure. But that would mean generating a custom collection for each product and locale pair. Might not be too difficult to do a nested loop though.

```js
  // .eleventy.js snippet
  for (const category of ["products", "articles", "campaigns", "reviews"]) {
    for (const locale of ["de", "en", "es", "fr"]) {
      const key = `${category}_${locale}`;
      eleventyConfig.addCollection(key, collectionApi => collectionApi.getFilteredByTag(key));
    }
  }
```

I can think of a few obvious potential bugs in my approach though (what if not all categories support all locales, etc). You could further refine by creating some map or object which might be more elegant &mdash; if you don't expect all locales to have the same content).

```js
const collectionMap = [
  {category: "products", locales: ["de", "en"]},
  {category: "articles", locales: ["en", "fr"]},
  ...
];
```

**Reply by pdehaan** [reply]
Seemed to work&hellip;

```js
  const collectionMap = [
    {category: "products", locales: ["de", "en"]},
    {category: "cats", locales: ["es", "fr"]}
  ];

  for (const {category, locales} of collectionMap) {
    for (const locale of locales) {
      const key = `${category}_${locale}`;
      eleventyConfig.addCollection(key, collectionApi => collectionApi.getFilteredByTag(key));
    }
  }
```

```nunjucks
{{ collections | keys | inspect | safe }}
```

### OUTPUT
```js
[ 'all', 'cats_es', 'cats_fr', 'products_de', 'products_en' ]
```

I only have one template in the `products_de` collection, so the rest of my collections should be empty. :shrug:

**Comment by madsem** [comment]
This "kinda" works haha :)

I am a little bit lost myself now, what you came up with gives us the desired result of multi-lingual collections.
But it seems I still would need to somehow create the tags dynamically (Which wasn't possible).

To elaborate, I added the following to my `.eleventy.js` just for testing:
```js
const locales = ["de", "en", "es", "fr"];
  const collectionMap = [
    {category: "products", locales: locales},
    {category: "articles", locales: locales},
    {category: "reviews", locales: locales}
  ];

  for (const {category, locales} of collectionMap) {
    for (const locale of locales) {
      const key = `${category}_${locale}`;
      config.addCollection(key, collectionApi => collectionApi.getFilteredByTag(key));
    }
  }
```

Then added this to my `home.njk` which is supposed to display all kinds of items from collections, in the language the visitor is currently viewing:
```nunjucks
       {{ collections | keys | inspect | safe }}
        {% for lander in collections["products_" + locale] %}
        <strong>{{lander.data.title}}</strong>
        {% endfor %}
```

Then I removed the `tags` property from all my directory data files like `products.11tydata.js`:
```js
module.exports = {
  layout: "page",
  permalink: "{{ locale }}/{{ slug or page.fileSlug }}/",
};
```

And the result displayed on my homepage was just this:
```log
[ 'all', 'articles_de', 'articles_en', 'articles_es', 'articles_fr', 'pages', 'products_de', 'products_en', 'products_es', 'products_fr', 'reviews_de', 'reviews_en', 'reviews_es', 'reviews_fr' ]
```

I then tried adding the `tags` property back, with a hardcoded locale:
```js
module.exports = {
  layout: "page",
  tags: ["products_de"],
  permalink: "{{ locale }}/{{ slug or page.fileSlug }}/",
};
```

And this time the actual output on the **German**  homepage was what I would expect (Minus the fact that now obvs. all collections are hardcoded to be "products_de").
The homepages of other locales, were only showing the debug output, not any of the collections (Which again, is the expected result in this test case)
```html
<div class="content flow">[ 'all', 'articles_de', 'articles_en', 'articles_es', 'articles_fr', 'pages', 'products_de', 'products_en', 'products_es', 'products_fr', 'reviews_de', 'reviews_en', 'reviews_es', 'reviews_fr' ]
 <strong>Cool List</strong> 
<strong>Coole Liste</strong>
</div>
```

I'm totally confused now, I thought that 11ty creates Collections one of two ways:
* By adding tags in frontmatter or datafile(s)
* By using `addCollection()` method.

Now, we used the `addCollection()` method, I'd expected it to already work, without adding tags...
But now it does work only if I ALSO add tags, and then also only if I hardcode the tags with locale.

I'm totally in love with 11ty, but it seems there is still a lot I don't know.

**Reply by madsem** [reply]
And PS:

On the site I am creating, all content/collections/pages etc support all languages we've used here.
It is possible though that the content is different on some pages, or that some locales don't have content that another locale has.

**Comment by madsem** [comment]
Ahh, sorry! I re-read what you wrote before, and realized you already posted the answer to what confused me :)
Dynamically created tags do not create collections for some reason, so we have to A: dynamically create tags using `eleventyComputed`, and B: add those to a collection.

I did exactly that now, and low and behold: All homepages display the content in their own language/or are empty if there is no content for that language.

Perfect :) Thanks so much for your help.

I'm curious if this will all just work now when I start adding localized categories.

**Reply by pdehaan** [reply]
Excellent! 🎉 

If all locales have the same content, we could probably tweak this code to use two simple arrays instead of the array of objects:

```js
  const locales = ["de", "en", "es", "fr"];
  const categories = ["products", "articles", "reviews"];

  for (const locale of locales) {
    // Add collections for "products_de", "articles_de", "reviews_de", "products_en", ...
    for (const category of categories) {
      const key = `${category}_${locale}`;
      eleventyConfig.addCollection(key, collectionApi => collectionApi.getFilteredByTag(key));
    }

    // Bonus collections for "all_de", "all_en", etc.
    // Might be useful if you want locale specific sitemaps or whatever.
    // But might not be useful if you have locale specific content outside the 3 categories which 
    // wouldn't be included in this collection. Although if it has a tag, we could easily add it
    // to this collection.
    eleventyConfig.addCollection(`all_${locale}`, collectionApi => {
      const localeTags = categories.map(category => `${category}_${locale}`);
      return collectionApi.getFilteredByTags(...localeTags);
    });
  }
```

Now if I add this code to my test.njk file:

```njk
{{ collections | keys | inspect | safe }}
```

### OUTPUT

```js
[
  'all',         'all_de',
  'all_en',      'all_es',
  'all_fr',      'articles_de',
  'articles_en', 'articles_es',
  'articles_fr', 'products_de',
  'products_en', 'products_es',
  'products_fr', 'reviews_de',
  'reviews_en',  'reviews_es',
  'reviews_fr'
]
```

**Reply by pdehaan** [reply]
There might be better solutions too, but I'd probably have to see your repo.
Not sure if there would be a way to add more directory data files somewhere so you can hardcode tags and have the collections automatically created. But maybe not if the pages are all created via pagination in a layout file.

**Reply by madsem** [reply]
Sure you can :) Invited you, it's private as it's supposed to be company internal for now.

The localized all collections are really interesting. I just really understood how powerful collections can be. Basically no limits in how content is structured.

**Reply by pdehaan** [reply]
Totally! Check out https://www.11ty.dev/docs/collections/#advanced-custom-filtering-and-sorting. Lots of great tips in there.
- `collectionApi.getAll()` &mdash; Returns an array.
- `collectionApi.getAllSorted()` &mdash; Use the default sorting algorithm (ascending by date, filename tiebreaker). Returns an array. 
- `collectionApi.getFilteredByTag( tagName )` &mdash; Returns an array.
- `collectionApi.getFilteredByTags( tagName, secondTagName, […] )` &mdash; Retrieve content that includes all of the tags passed in. Returns an array.
- `collectionApi.getFilteredByGlob( glob )` &mdash; Will match an arbitrary glob (or an array of globs) against the input file’s full `inputPath` (including the input directory). Returns an array.

Plus you can do additional filtering before creating the collection. So you could use `.getAllSorted()` to get all the content, and then inspect each template's frontmatter and only add it to a collection if it has `locale: fr`. Although I think we've had issues trying to create dynamic collections while currently trying to add a collection.
Initially I was thinking we could use `.getFilterdByGlob()` but that might be tricky since your "de" templates could be in multiple different folders. And I'd have to see if something like `getFilteredByGlob("**/de/**")` works.

**Reply by madsem** [reply]
Categories are supposed to be a little different than what's regularly used in blogs for example.

Products can be assigned topics from a short list of entries in a data file (Which on the front end also needs localization)

Then categories/topics are used on listing type pages to display products.

Products then can be related to reviews and articles, which are supposed to be displayed on a products detail page as helpful content.

Does this sound reasonable or does 11ty pagination favtor in here in any way that would make more sense then the above?

**Reply by pdehaan** [reply]
No, I think that makes sense...

Looking at your repo, I'm wondering if you could just add `tags` to your /site/authors/de/de.json data file (instead of using `eleventyComputed`=>`tags` in your /site/authors/authors.11tydata.js file).

```diff
{
+ "tags": ["authors_de"],
  "dir": "ltr",
  "locale": "de"
}
```

```diff
module.exports = {
  layout: "page",
  permalink: "{{ locale }}/authors/{{ page.fileSlug }}/",
- eleventyComputed: {
-   tags(data) {
-     return [`authors_${ data.locale }`];
-   }
- }
};
```

Sure, it's a bit of duplication since you need to put it in each locale's folder, and make sure it matches the defined `locale` property as well. But I think that should auto-create all the collections for you.
I'd have to try creating a smaller test case to see if you could use an de.11tydata.js file which creates the tag name based on the specified locale, or if that'd just break the auto-collection creation anyways.

**Reply by pdehaan** [reply]
Meh, I don't know if I like it.

```js
// /site/products/de/de.11tydata.js
module.exports = () => {
  const locale = "de";
  return {
    locale,
    dir: "ltr",
    tags: [`products_${locale}`]
  };
};
```

Also unclear on if you need `dir` defined there. It seems like the locale's text direction is constant. So instead of defining it in 4+ places throughout the repo (per locale). It could possibly be defined once in a global data file.

```js
// /site/_data/locale_dir.js
module.exports = {
  "de": "ltr",
  "en": "ltr",
  "es": "ltr",
  "fr": "ltr"
};
```

Now, as long as your `locale` is defined somewhere, you could probably look up the text direction for your layout via `locale_dir[locale]`.

**Reply by pdehaan** [reply]
Or possibly even https://www.npmjs.com/package/rtl-detect

You could see if you can wrap it in a custom filter.

```js
const {getLangDir} = require("rtl-detect");

module.exports = eleventyConfig => {
  eleventyConfig.addFilter("dir_from_locale", (locale) => {
    return getLangDir(locale) || "ltr";
  });

  // Or just this one liner:
  // eleventyConfig.addFilter("dir_from_locale", getLangDir);
  ...
};
```

```njk
en={{ "en" | dir_from_locale }}
ja={{ "ja" | dir_from_locale }}
he={{ "he" | dir_from_locale }}
ar={{ "ar" | dir_from_locale }}
```

### OUTPUT

```ini
en=ltr
ja=ltr
he=ltr
ar=rtl
```

Oddly, I thought Hebrew (he) was "rtl" ([1](https://github.com/shadiabuhilal/rtl-detect/blob/2eed8a33276461a24e7033d1d3a115ee64aee3f5/lib/rtl-detect.js#L133-L161)), so that seems a bit unexpected.

**Reply by pdehaan** [reply]
> Oddly, I thought Hebrew (he) was "rtl" ([1](https://github.com/shadiabuhilal/rtl-detect/blob/2eed8a33276461a24e7033d1d3a115ee64aee3f5/lib/rtl-detect.js#L133-L161)), so that seems a bit unexpected.

Blah... Looks like Hebrew/Yiddish worked in rtl-detect@1.0.2 but broke in v1.0.3. Filed upstream as https://github.com/shadiabuhilal/rtl-detect/issues/11.

**Reply by madsem** [reply]
Ohhh, I see. That is actually pretty cool 😎 

Today I wanted to start implementing the layouts, perfect time to test if I can maybe use those locale data files.

I assume we probably, most likely won't ever write in a language other than rtl. But you never know! So it's definitely a good idea to have this all configured as central as possible.

Which is also what I've tried to implement with the `_data/site.json` that holds all the settings, including the languages.

The end goal is to be able to add languages easily just by adding them to the site.json and copy the locale folders for all collections.

Although it's not clear yet if we will ever need to add more. I just wanted to make sure that we can if that need ever comes up. 

The entire template as currently planned, is going to be used for only 3 different websites, which are all medium sized product review sites. Think maybe 500 product comparison pages where each product has approximately 5-10 helpful reviews and articles.

And once it's done, I wanted to take some time to make a theme out of it. Just take the company internal / opinionated stuff out and make it available as 11ty theme. Because there aren't many multi lingual ones yet :)

Will now see if i can use the same methods you showed me for the `_data/topics.json`. Which, I think, needs to be moved into the i18n dir and also translated, and then also added as collection in the CMS, so it can be selected by the writers.

**Reply by madsem** [reply]
It's a little more complex than I initially thought.
I got the categories to work in a sense that in the CMS they are selectable from a `select` widget.

```js
// site/_data/topics.js
module.exports = [
  "Lifestyle",
  "Business",
 ...
];
```
Then I created a netlify cms field:
```js
import topicsList from "../../_data/topics.js";
{
      i18n: "duplicate",
      label: "Select a Topic",
      name: "topic",
      widget: "select",
      multiple: false,
      hint: "Select Landing Page Topic. Topics act as categories.",
      options: topicsList,
    },
```

This is probably as good as it gets in terms of localizing the CMS UI. I guess writers can select it in english, no matter which language they write in, and then during built time it can be localized based on the english name. (Similar to how you suggested to do it with the locale dir.

Now this creates front-matter with topics assigned like:
```yaml
---
topic: "Lifestyle"
---
```

And on the template side of things, I can now easily loop through the data file
```nunjucks
{% for topic in topics %}
<li><a href="/{{ topic | slug }}/" class="box-display-block">{{ topic }}</a></li>
{% endfor %}
```

The problematic parts come now haha
All other urls on the site are locale aware (kinda), the urls built by 11ty include the locale as the locale dirs are part of everything.

The topics on the other hand, are not.
Sure, one could now simply do `{{ locale }}/{{ topic | slug }}/`, but that feels wrong and extremely brittle in the long run.

Is there a way to handle the url management for all these things in a central way?
Maybe by using a filter like `{{ topic | i18n_url }}`, maybe all this falls in the same category as localization of theme strings (Which is next on my todo list)

**Reply by madsem** [reply]
Actually came up with this, but not sure if that will cause issues, hmmm.

```nunjucks
 <li><a href="/{{ topic | slug | i18n_url(locale) }}" class="mt-300 mb-300 color-primary weight-regular hover-color-secondary box-block">{{ topic }}</a></li>

```

```js
// .eleventy.js
/**
   * Localized url helper
   * usage: {{ item | slug | i18n_url(locale) }}
   */
  config.addFilter("i18n_url", function(url, locale) {
    return locale + "/" + url
  });
```

Although urls are now including their locale, there's still also the issue of displaying these urls in the form of listing pages haha

**Reply by pdehaan** [reply]
> ```js
>   config.addFilter("i18n_url", function(url, locale) {
>     return locale + "/" + url
>   });
> ```

Nice!
A couple quick suggestions:
1. Probably safest to prefix that with a `/` so the links are all absolute instead of relative (in case pages live at different depths, you wouldn't oddly end up with "/de/de/cool-list".
    ```js
    return `/${locale}/${url}`;
    ```
    Or maybe not. After re-re-reading your `a href` I see a `/` prefix before the `{{topic|...}}`. Although it might be pretty easy to forget to add that `/` on a page, and it's safer to keep the logic in the custom filter. 🤷 

2. Not sure if you want a default value for the `locale` attribute (or throw some error on a missing `locale` attribute). Currently if you forget to specify the locale, it will almost certainly be a 404:
    ```njk
    {{ topic | slug | i18n_url }}"
    ```
    Since that would link to something like `/privacy-policy` (or `//privacy-policy` if you do the leading backslash suggestion above).
    Possibly safest to default to English, although that could be weird if you're on a German page and suddenly the site switches locales because somebody forgot to add `(locale)` at the end of the `i18n_url` filter.
    ```diff
    -config.addFilter("i18n_url", function(url, locale) {
    +config.addFilter("i18n_url", function(url, locale="en") {
      return locale + "/" + url
    });
    ```
    I guess it's a balance of better to 404 or better to switch locales. On one hand, 404s are a terrible user experience, but easier to catch with a link checking tool. Whereas switching locales is arguably the better of two poor user experiences but much more difficult to catch with a link checker. We could _not_ add a default `locale` attribute value, display an error message if the `locale` is `undefined` and then `return (locale || "en") + "/" + url;` to default it to English anyways, but that might be unhelpful too since I don't think filters have any context of which page/template they're being run on, so you'd be console logging "Missing 'locale' for ${url}. Defaulting to '/en/'." and then you have to go through each of your templates to find where you called the `i18n_url` filter without a locale.

3. Wondering if there is a way to call the `slug` filter from within the `i18n_url` filter automatically so you have less filter chaining. Might be as easy as:
    ```js
    config.addFilter("i18n_url", function(url, locale="en") {
      const slugFn = config.getFilter("slug");
      return `/${locale}/${slugFn(url)}`;
    });
    ```
    But that might somewhat simplify your usage to `<a href="{{ topic | i18n_url(locale) }}"`

**Reply by madsem** [reply]
And translations can now easily be done like
```nunjucks
 {% for topic in topics[locale] %}
                <li><a href="/{{ topic | slug | i18n_url(locale) }}" class="mt-300 mb-300 color-primary weight-regular hover-color-secondary box-block">{{ topic }}</a></li>
                {% endfor %}
```

I would just have to adjust the data file for topics, and would then even be able to have internationalized url slugs.
But what I'm still entirely in the dark about, is how to display the content that is assigned to a topic.

I assume I will need to create a file for that, but how would that be connected with the topics data file.
I think I might've forgotten some very important step here haha, or better: I don't know one important step here.

**Reply by madsem** [reply]
Yes, thanks! Your version is much better :)
I was thinking about these forward slashes in the theme as well, your way is better.

**Reply by madsem** [reply]
My head probably doesn't work so well anymore at the end of my day over here haha, but isn't the solution to displaying content under certain topics, to create a collection in the eleventy configuration.

Then create a nunjucks template file with collection: topics

And there i can simply loop and paginate over content for each topic.

So in the end there's one page for every topic, with a list of content?

**Reply by madsem** [reply]
actually, just got word that internationalization is postponed and i just have to integrate US english haha jesus.

Thanks for all the help @pdehaan if you're ever in Germany or Houston drop me a line, would love to invite you to beers 🍻 :) Or non-alcoholic beverages of course also!

I think I will continue to play around with the i18n version though and release it as a theme.
